tempodb/blocklist/list.go

```diff
@@ -125,64 +125,70 @@ func (l *List) Update(tenantID string, add []*backend.BlockMeta, remove []*backe
 // updateInternal exists to do the work of applying updates to held PerTenant and PerTenantCompacted maps
 // it must be called under lock
 func (l *List) updateInternal(tenantID string, add []*backend.BlockMeta, remove []*backend.BlockMeta, compactedAdd []*backend.CompactedBlockMeta, compactedRemove []*backend.CompactedBlockMeta) {
-	// ******** Regular blocks ********
-	blocklist := l.metas[tenantID]
-
-	matchedRemovals := make(map[uuid.UUID]struct{})
-	for _, b := range blocklist {
-		for _, rem := range remove {
-			if b.BlockID == rem.BlockID {
-				matchedRemovals[(uuid.UUID)(rem.BlockID)] = struct{}{}
-				break
-			}
+	hasID := func(id backend.UUID) func(*backend.BlockMeta) bool {
+		return func(b *backend.BlockMeta) bool {
+			return b.BlockID == id
 		}
 	}
 
-	existingMetas := make(map[uuid.UUID]struct{})
-	newblocklist := make([]*backend.BlockMeta, 0, len(blocklist)-len(matchedRemovals)+len(add))
-	// rebuild the blocklist dropping all removals
-	for _, b := range blocklist {
-		existingMetas[(uuid.UUID)(b.BlockID)] = struct{}{}
-		if _, ok := matchedRemovals[(uuid.UUID)(b.BlockID)]; !ok {
-			newblocklist = append(newblocklist, b)
+	hasIDC := func(id backend.UUID) func(*backend.CompactedBlockMeta) bool {
+		return func(b *backend.CompactedBlockMeta) bool {
+			return b.BlockID == id
 		}
 	}
-	// add new blocks (only if they don't already exist)
-	for _, b := range add {
-		if _, ok := existingMetas[(uuid.UUID)(b.BlockID)]; !ok {
-			newblocklist = append(newblocklist, b)
-		}
-	}
-
-	l.metas[tenantID] = newblocklist
 
-	// ******** Compacted blocks ********
-	compactedBlocklist := l.compactedMetas[tenantID]
-
-	compactedRemovals := map[uuid.UUID]struct{}{}
-	for _, c := range compactedBlocklist {
-		for _, rem := range compactedRemove {
-			if c.BlockID == rem.BlockID {
-				compactedRemovals[(uuid.UUID)(rem.BlockID)] = struct{}{}
-				break
+	// ******** Regular blocks ********
+	if len(add) > 0 || len(remove) > 0 || len(compactedAdd) > 0 || len(compactedRemove) > 0 {
+		var (
+			existing = l.metas[tenantID]
+			final    = make([]*backend.BlockMeta, 0, max(0, len(existing)+len(add)-len(remove)))
+		)
+
+		// rebuild dropping all removals
+		for _, b := range existing {
+			if slices.ContainsFunc(remove, hasID(b.BlockID)) {
+				continue
 			}
+			final = append(final, b)
 		}
-	}
+		// add new if they don't already exist and weren't also removed
+		for _, b := range add {
+			if slices.ContainsFunc(final, hasID(b.BlockID)) ||
```

Comment:
before we built a map for lookups and it looks like we're just doing linear searches here. wdyt think about sorting and doing a binary search?

Reply:
Sorry I forgot to include the benchmarks but the this version is already significantly faster, mainly by eliminating the maps, which reduces allocs.  

```
pkg: github.com/grafana/tempo/tempodb/blocklist
cpu: Apple M4 Pro
       │ before.txt  │             after.txt              │
       │   sec/op    │   sec/op     vs base               │
Update   5.485m ± 2%   1.212m ± 2%  -77.91% (p=0.002 n=6)

       │  before.txt  │              after.txt              │
       │     B/op     │     B/op      vs base               │
Update   6.345Mi ± 0%   1.746Mi ± 0%  -72.48% (p=0.002 n=6)

       │  before.txt   │             after.txt             │
       │   allocs/op   │ allocs/op   vs base               │
Update   4079.000 ± 0%   4.000 ± 0%  -99.90% (p=0.002 n=6)
```

Since the number of metas to add/remove is low (max 4?), sorting/binsearch may not be much better linear, but I can check.

Reply:
oh, ok. This is called with just a few metas each compaction to pass the results of the cycle into the polled blocklist. And then each polling cycle it's called once with all accumulated modifications that cycle b/c it assumes that the tenant index is still a bit behind.

given everything the compactor does this may not be noticeable with either a linear or binary search. i'm good as is and will approve. up to you if you want to explore binary search.

Reply:
Tested a bit and sorting the current blocklist is outweighing the benefits of the quicker lookups (benchmark using 100K blocklist).  I checked and the even the current implementation isn't showing up in profiles, so I think it's not worth digging more in here.